### PR TITLE
Fixed wrong uri paths from subresources

### PR DESCRIPTION
**Pull request by AndreasBieber**
Hi Basti,

when using subresources from a specific element/resource item (i.e. `/resources/{id}/subresource`) the `{id}` will not be part of the whole uri (i.e. `/resources/subresource`).

So this pullrequest fixes this bug by just calling `parent.EnsureTrailingSlash().`.

I've also provided a unittest to proove this fix.

Cheers,
Andy

### CODE REVIEW

**Comment by bastianeicher** [comment]
Hi Andy,

"Its a feature, not a bug!" ;)

Jokes aside: Currently only the constructor for `CollectionEndpointBase` calls `.EnsureTrailingSlash()` because collections are always expected to have children. The reason other endpoint types don't do this as well, is that the implementation of the REST API (if it is something other than Web API) might actually expect an URI without a trailing slash.

Two possible alternative solutions:
 * Deriving your own custom `SubresourceEndpoint` from `ElementEndpoint` allows you to add `.EnsureTrailingSlash()` there.
 * Using HTTP `Link` headers enables relative URIs to resolved by the server side.

I'm definitely open to adding `.EnsureTrailingSlash()` to more places in TypedRest, but we'd have to make sure it remains configurable enough for all scenarios. Any thoughts on this?

Cheers
Bastian

**Comment by AndreasBieber** [comment]
Hi Basti,

i don't know if you misunderstood me or i misunderstood you.

The problem is not the missing trailing slash at the `relativeUri` as it'll be appended in `CollectionEndpointBase`. The problem is the missing trailing slash at the `parent.Uri`.

**An example for better understanding**
I have the following scenario:
`/api/template/{id}` <- this will return an `ElementEndpoint` which has the following property:
`public TemplateDetailEndpoint Detail => new TemplateDetailEndpoint(this, "detailcontent");` which should result in the uri `/api/template/{id}/detailcontent` right?

Now, when i'll access `ElementEndpoint.Detail.Uri` i would expect the following uri: `/api/template/{id}/detailcontent` but i'm getting `/api/template/detailcontent`.

As i'm only the consumer from an api, the second alternative solution cannot be applied.
The first solution seems to be over the top just for having a trailing slash at the `parent.Uri`.

I understand your objection makes sense to handover the decision to the consumer. But as i already mentioned, the approach to create a custom ElementPoint seems not to be the right solution IMHO.

What do you think about to extend the existing ctors inside of `EndpointBase `like this:
```
protected EndpointBase(IEndpoint parent, Uri relativeUri, bool ensureTrailingSlashOnParentUri = false)
    : this(
        parent.HttpClient,
        new Uri(ensureTrailingSlashOnParentUri ? parent.Uri.EnsureTrailingSlash() : parent.Uri, relativeUri))
{
}
```
```
protected EndpointBase(IEndpoint parent, string relativeUri, bool ensureTrailingSlashOnParentUri = false)
    : this(parent, new Uri(relativeUri, UriKind.Relative), ensureTrailingSlashOnParentUri)
{
}
```

Cheers,
 Andy

**Comment by bastianeicher** [comment]
Since you are adding a custom property to the endpoint representing `/api/template/{id}`, namely `Detail`, you must already be using a custom class derived from `ElementEndpoint<T>` here, right? So this class could look something like this:
```csharp
public class TemplateEndpoint : ElementEndpoint<Template>
{
    public TemplateEndpoint(IEndpoint parent, Uri relativeUri)
        : base(parent, relativeUri.EnsureTrailingSlash())
    {}

    public TemplateDetailEndpoint Detail => new TemplateDetailEndpoint(this, "detailcontent");
}
```
Now `/api/template/{id}` would always turn into `/api/template/{id}/`, causing the `detailcontent` relative link to correctly resolve to `/api/template/{id}/detailcontent`.

Basically, I prefer appending the trailing slash right to the containing endpoint rather than adding it only for resolving relative URIs. This makes the relative URIs behave just like relative links on an ordinary HTML page.

However, if you have a scenario where this does not work or makes the code bloated, the `bool ensureTrailingSlashOnParentUri` optional constructor parameter sounds like a good solution.

**Comment by AndreasBieber** [comment]
Hi Basti,

that's a good idea and it should solve **my** problem. But anyway we should implement this optional parameter, because with your solution the consumer has to decide whether he calls 

/api/template/{id} or /api/template/{id}**/**

So the api i'm currently using supports the trailing slash after the `{id}`, but as you already mentioned, what if the api does'nt support the trailing slash after the `{id}` but is providing a subresource?

Can i implement the optional parameter or will do you?

What do you think? 

Cheers
 Andy

**Comment by bastianeicher** [comment]
Hi Andy,

sure, you can implement the optional parameter, thanks. :)
I'll close this pull request and you can open a new one with this change.

Regards
Bastian
